From 180737e1b98fe58677c1447a1ff2f034bad62c7e Mon Sep 17 00:00:00 2001
From: Lars Schmertmann <Lars.Schmertmann@governikus.de>
Date: Mon, 20 Oct 2025 10:43:52 +0200
Subject: [PATCH 546/553] Fix native a11y role of QAccessible::LayeredPane

* The QML StackView started to use this role in
  qtdeclarative ee7151d5ab29ba812a93c1231ddc59c9129a7af0.
* This will replace "Unknown" with "Window" on Windows with NVDA.
* On macOS the behavior with VoiceOver is not changed but
  it seems sensible to supplement the mapping as well.

Fixes: QTBUG-134138
Pick-to: 6.8
Change-Id: Iab5c7339a82d91b6eafdba45fa31c0537a94a135
Reviewed-by: Michael Weghorn <m.weghorn@posteo.de>
(cherry picked from commit 6b4da1662fa5688d489f75aba76bf15e707448f8)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/plugins/platforms/cocoa/qcocoaaccessibility.mm              | 1 +
 src/plugins/platforms/windows/uiautomation/qwindowsuiautils.cpp | 1 +
 2 files changed, 2 insertions(+)

diff --git a/src/plugins/platforms/cocoa/qcocoaaccessibility.mm b/src/plugins/platforms/cocoa/qcocoaaccessibility.mm
index c825919e38e..207c28398bf 100644
--- a/src/plugins/platforms/cocoa/qcocoaaccessibility.mm
+++ b/src/plugins/platforms/cocoa/qcocoaaccessibility.mm
@@ -159,6 +159,7 @@ static void populateRoleMap()
     roleMap[QAccessible::Graphic] = NSAccessibilityImageRole;
     roleMap[QAccessible::Tree] = NSAccessibilityOutlineRole;
     roleMap[QAccessible::BlockQuote] = NSAccessibilityGroupRole;
+    roleMap[QAccessible::LayeredPane] = NSAccessibilityGroupRole;
 }
 
 /*
diff --git a/src/plugins/platforms/windows/uiautomation/qwindowsuiautils.cpp b/src/plugins/platforms/windows/uiautomation/qwindowsuiautils.cpp
index 7536ece4dbc..780bdfab533 100644
--- a/src/plugins/platforms/windows/uiautomation/qwindowsuiautils.cpp
+++ b/src/plugins/platforms/windows/uiautomation/qwindowsuiautils.cpp
@@ -151,6 +151,7 @@ long roleToControlTypeId(QAccessible::Role role)
         {QAccessible::WebDocument, UIA_DocumentControlTypeId},
         {QAccessible::Heading, UIA_TextControlTypeId},
         {QAccessible::BlockQuote, UIA_GroupControlTypeId},
+        {QAccessible::LayeredPane, UIA_PaneControlTypeId},
     };
 
     return mapping.value(role, UIA_CustomControlTypeId);
-- 
2.51.2

